AppTaskQueue/appscale/taskqueue/queue.py

```diff
@@ -235,7 +235,538 @@ def __repr__(self):
+    if not hasattr(task, 'payloadBase64'):
+      raise InvalidTaskInfo('{} is missing a payload.'.format(task))
+
+    enqueue_time = datetime.datetime.utcnow()
+    try:
+      lease_expires = task.leaseTimestamp
+    except AttributeError:
+      lease_expires = datetime.datetime.utcfromtimestamp(0)
+
+    try:
+      task.queueName = self.name
+      task.enqueueTimestamp = enqueue_time
+      task.leaseTimestamp = lease_expires
+
+      self.pg_cursor.execute(
+        'INSERT INTO "{table}" ( '
+        '  task_name, payload, time_enqueued, '
+        '  lease_expires, retry_count, tag '
+        ')'
+        'VALUES ( '
+        '  %(task_name)s, %(payload)s, %(time_enqueued)s, '
+        '  %(lease_expires)s, %(retry_count)s, %(tag)s '
+        ')'.format(table=self.tasks_table_name),
+        vars={
+          'task_name': task.id,
+          'payload': task.payloadBase64,
+          'time_enqueued': enqueue_time,
+          'lease_expires': lease_expires,
+          'retry_count': 0,
+          'tag': getattr(task, 'tag', None)
+        }
+      )
+
+      self.pg_connection.commit()
+      logger.debug('Added task: {}'.format(task))
+
+    except psycopg2.IntegrityError as err:
+      name_taken_msg = 'Task name already taken: {}'.format(task.id)
+      logger.warning('{msg}. Rolling back transaction({err})'
+                     .format(msg=name_taken_msg, err=err))
+      self.pg_connection.rollback()
+      raise InvalidTaskInfo(name_taken_msg)
+    except Exception as err:
+      logger.error('Rolling back transaction ({err})'.format(err=err))
+      self.pg_connection.rollback()
+      raise
+
+  def get_task(self, task, omit_payload=False):
+    """ Gets a task from the queue.
+
+    Args:
+      task: A Task object.
+      omit_payload: A boolean indicating that the payload should not be
+        fetched.
+    Returns:
+      A task object or None.
+    """
+    if omit_payload:
+      columns = ['task_name', 'time_enqueued',
+                 'lease_expires', 'retry_count', 'tag']
+    else:
+      columns = ['payload', 'task_name', 'time_enqueued',
+                 'lease_expires', 'retry_count', 'tag']
+    try:
+      self.pg_cursor.execute(
+        'SELECT {columns} FROM "{tasks_table}" '
+        'WHERE task_name = %(task_name)s'
+        .format(columns=', '.join(columns),
+                tasks_table=self.tasks_table_name),
+        vars={
+          'task_name': task.id,
+        }
+      )
+      row = self.pg_cursor.fetchone()
+      self.pg_connection.commit()
+    except Exception as err:
+      logger.error('Rolling back transaction ({err})'.format(err=err))
+      self.pg_connection.rollback()
+      raise
+
+    if not row:
+      return None
+    return self._task_from_row(columns, row, id=task.id)
+
+  def delete_task(self, task):
+    """ Deletes a task from the queue.
+
+    Args:
+      task: A Task object.
+    """
+    try:
+      self.pg_cursor.execute(
+        'DELETE FROM "{tasks_table}" '
```

Comment:
We may want to delete all the fields (except the task_name and maybe a field to mark when it was deleted) instead of deleting the whole row in order to tombstone the task name. A periodic task could then clear all names that were tombstoned earlier than a set interval into the past.